src/shogun/base/SGObject.h

```diff
@@ -1022,5 +1029,123 @@ class CSGObject
 	/** Subscriber used to call onNext, onComplete etc.*/
 	SGSubscriber* m_subscriber_params;
 };
+
+	class ToStringVisitor : public AnyVisitor
+	{
+	public:
+		ToStringVisitor(std::stringstream* ss) : AnyVisitor(), m_stream(ss)
+		{
+		}
+
+		virtual void on(bool* v)
+		{
+			stream() << (*v ? "true" : "false");
+		}
+		virtual void on(int32_t* v)
+		{
+			stream() << *v;
+		}
+		virtual void on(int64_t* v)
+		{
+			stream() << *v;
+		}
+		virtual void on(float* v)
+		{
+			stream() << *v;
+		}
+		virtual void on(double* v)
+		{
+			stream() << *v;
+		}
+		virtual void on(long double* v)
+		{
+			stream() << *v;
+		}
+		virtual void on(CSGObject** v)
+		{
+			if (*v)
+			{
+				stream() << (*v)->get_name() << "(...)";
+			}
+			else
+			{
+				stream() << "null";
+			}
+		}
+		virtual void on(SGVector<int>* v)
+		{
+			to_string(v);
+		}
+		virtual void on(SGVector<float>* v)
+		{
+			to_string(v);
+		}
+		virtual void on(SGVector<double>* v)
+		{
+			to_string(v);
+		}
+		virtual void on(SGMatrix<int>* mat)
+		{
+			to_string(mat);
+		}
+		virtual void on(SGMatrix<float>* mat)
+		{
+			to_string(mat);
+		}
+		virtual void on(SGMatrix<double>* mat)
+		{
+			to_string(mat);
+		}
+
+	private:
+		std::stringstream& stream()
+		{
+			return *m_stream;
+		}
+
+		template <class T>
+		void to_string(SGMatrix<T>* m)
+		{
+			if (m)
+			{
+				stream() << "Matrix<" << demangled_type<T>() << ">(" << m->num_rows
+						 << "," << m->num_cols << "): [";
+				for (auto col : range(m->num_cols))
+				{
+					stream() << "[";
+					for (auto row : range(m->num_rows))
+					{
+						stream() << (*m)(row, col);
+						if (row < m->num_rows - 1)
+							stream() << ",";
+					}
+					stream() << "]";
+					if (col < m->num_cols)
+						stream() << ",";
+				}
+				stream() << "]";
+			}
+		}
+
+		template <class T>
+		void to_string(SGVector<T>* v)
+		{
+			if (v)
+			{
+				stream() << "Vector<" << demangled_type<T>() << ">(" << v->vlen
+						 << "): [";
+				for (auto i : range(v->vlen))
+				{
+					stream() << (*v)[i];
+					if (i < v->vlen - 1)
+						stream() << ",";
+				}
+				stream() << "]";
+			}
+		}
+
+	private:
+		std::stringstream* m_stream;
+	};
```

Comment:
Reminder to self, once #4594 is done can put this back into place and use SGVector to_string method

Reply:
merged :)